keras/backend/common/dtypes.py

```diff
@@ -0,0 +1,297 @@
+            f"{extended_allowed_dtypes}. Received: dtype={dtype}"
+        )
+    if precision not in ["16", "32", "64"]:
+        raise ValueError(
+            f"Invalid value for argument `precision`. Expected one of "
+            f"('16', '32', '64'). Received: precision={precision}"
+        )
+    if dtype == "bfloat16":  # special case for bfloat16
+        dtype_indicator = "f"
+    else:
+        dtype_indicator = dtype[:1]
+
+    if dtype_indicator == "b":
+        return "bool"
+    elif dtype_indicator == "i":
+        return "int" + precision
+    elif dtype_indicator == "u":
+        return "uint" + precision
+    else:
+        return "float" + precision
+
+
+BIT64_TO_BIT16_DTYPE = {
+    "int32": "int16",
+    "int64": "int16",
+    "uint32": "uint16",
+    "uint64": "uint16",
+    "float32": "float16",
+    "float64": "float16",
+}
+BIT64_TO_BIT32_DTYPE = {
+    "int64": "int32",
+    "uint64": "uint32",
+    "float64": "float32",
+}
+
+
+@functools.lru_cache(maxsize=None)
+def _canonicalize_dtype_by_precision(dtype, precision="32"):
+    """Canonicalize dtype by the precision of `backend.floatx()`."""
+    if precision == "16":
+        return BIT64_TO_BIT16_DTYPE.get(dtype, dtype)
+    elif precision == "32":
+        return BIT64_TO_BIT32_DTYPE.get(dtype, dtype)
+    elif precision == "64":
+        return dtype
+    else:
+        raise ValueError(
+            f"Invalid value for argument `precision`. Expected one of "
+            f"('16', '32', '64'). Received: precision={precision}"
+        )
+
+
+def _lattice_result_type(*args):
+    dtypes, weak_types = zip(*(_dtype_and_weaktype(arg) for arg in args))
+    if len(dtypes) == 1:
+        out_dtype = dtypes[0]
+        out_weak_type = weak_types[0]
+    elif len(set(dtypes)) == 1 and not all(weak_types):
+        # Trivial promotion case. This allows extended dtypes through.
+        out_dtype = dtypes[0]
+        out_weak_type = False
+    elif all(weak_types):
+        # If all inputs are weakly typed, we compute the bound of the
+        # strongly-typed counterparts and apply the weak type at the end. This
+        # avoids returning the incorrect result with non-canonical weak types
+        # (e.g. weak int16).
+        out_dtype = _least_upper_bound(
+            *{_respect_weak_type(d, False) for d in dtypes}
+        )
+        out_weak_type = True
+    else:
+        nodes = []
+        for d, w in zip(dtypes, weak_types):
+            nodes.append(_respect_weak_type(d, w))
+        out_dtype = _least_upper_bound(*nodes)
+        out_weak_type = any(out_dtype is t for t in WEAK_TYPES)
+
+    out_weak_type = (out_dtype != "bool") and out_weak_type
+    if out_weak_type:
+        out_dtype = _resolve_weak_type(out_dtype)
+    precision = backend.floatx()[-2:]
+    out_dtype = _canonicalize_dtype_by_precision(out_dtype, precision)
+    return out_dtype
+
+
+@keras_export("keras.backend.result_dtype")
+def result_dtype(*tensors_and_dtypes):
```

Comment:
It's confusing that this takes both dtypes and tensors. Why not pick one, e.g. only pass actual tensors / floats / ints?

Reply:
Changed to `dtypes`, implying that we can only accept dtypes and types.